dhall/src/Dhall/Schemas.hs

```diff
@@ -0,0 +1,206 @@
+data Schemas = Schemas
+    { characterSet :: CharacterSet
+    , censor       :: Censor
+    , input        :: Input
+    , outputMode   :: OutputMode
+    , schemas      :: Text
+    }
+
+-- | Implementation of the @dhall schemas@ subcommand
+schemasCommand :: Schemas -> IO ()
+schemasCommand Schemas{..} = do
+    originalText <- case input of
+        InputFile file -> Text.IO.readFile file
+        StandardInput  -> Text.IO.getContents
+
+    (Header header, expression) <- Util.getExpressionAndHeaderFromStdinText censor originalText
+
+    schemasRecord <- Core.throws (Parser.exprFromText "(schemas)" schemas)
+
+    schemasExpression <- simplifyUsingSchemas schemasRecord expression
+
+    let docStream =
+            Dhall.Pretty.layout
+                (   Pretty.pretty header
+                <>  Dhall.Pretty.prettyCharacterSet characterSet schemasExpression
+                <>  "\n"
+                )
+
+    let schemasText = Pretty.Text.renderStrict docStream
+
+    case outputMode of
+        Write -> do
+            case input of
+                InputFile file -> do
+                    if originalText == schemasText
+                        then return ()
+                        else AtomicWrite.atomicWriteFile
+                                file
+                                (Pretty.Text.renderLazy docStream)
+                StandardInput -> do
+                    supportsANSI <- ANSI.hSupportsANSI IO.stdout
+
+                    Pretty.Terminal.renderIO
+                        IO.stdout
+                        (if supportsANSI
+                            then fmap Dhall.Pretty.annToAnsiStyle docStream
+                            else Pretty.unAnnotateS docStream)
+
+        Check -> do
+            if originalText == schemasText
+                then return ()
+                else do
+                    let command = "rewrite-with-schemas"
+
+                    let modified = "rewritten"
+
+                    Exception.throwIO CheckFailed{..}
+
+decodeSchema :: Expr Src Void -> Maybe (Expr Src Void, Map Text (Expr Src Void))
+decodeSchema (RecordLit m)
+        | Just  _Type               <- Map.lookup "Type" m
+        , Just (RecordLit _default) <- Map.lookup "default" m =
+            Just (_Type, _default)
+decodeSchema _ =
+    Nothing
+
+decodeSchemas
+    :: Expr Src Void
+    -> Maybe (Map SHA256Digest (Text, Map Text (Expr Src Void)))
+decodeSchemas (RecordLit keyValues) = do
+    m <- traverse decodeSchema keyValues
+
+    let typeMetadata = Map.fromList $ do
+            (name, (_Type, _default)) <- Map.toList m
+
+            return (Import.hashExpression (Syntax.denote _Type), (name, _default))
+
+    return typeMetadata
+decodeSchemas  _ = do
+    empty
+
+-- | Simplify a Dhall expression using a record of schemas
+simplifyUsingSchemas
+    :: Expr Src Import
+    -- ^ Record of schemas
+    -> Expr Src Import
+    -- ^ Expression to simplify using the supplied schemas
+    -> IO (Expr Src Import)
```

Comment:
This seems a bit more "honest".

```suggestion
    -> IO (Expr s a)
```

Reply:
The inferred type uses `Src`.  The reason why is because of the `let`-bound `schemas` in the rewritten expression, which preserves the original `Src` annotations

Reply:
Oh, right!